### PR TITLE
Refactor mage files

### PR DESCRIPTION
**Pull request by rvolosatovs**
#### Summary

References #2280 #2731 https://github.com/TheThingsNetwork/lorawan-stack/issues/2769

#### Changes

- Remove `js:devDeps`
- Install yarn and all dependencies on `js:deps`
- Do not use custom target check logic, rely on `target` package
- Properly set `mg.Deps` for JS targets
- Use `filepath.Join`
- Simplify JS-related functionality
- Do not call binaries directly - use `npx` and `yarn run`
- Do not attempt to run tools from different workspaces, e.g. avoid running `babel` from root in `sdk/js`, instead run `babel` from `sdk/js`
- Remove `jsSDK:link` - that is now part of `js:deps`.

#### Testing

CI, locally

```
$ rm -rf node_modules doc/themes/the-things-stack/node_modules sdk/js/node_modules && mage js:clean jsSDK:clean js:cleanDeps jsSDK:cleanDeps && mage <target>
```

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by rvolosatovs** [comment]
Hmm linting fails with:
```
exec: node_modules/.bin/eslint ./pkg/webui/**/*.js ./config/**/*.js --no-ignore --color --quiet

/home/travis/gopath/src/go.thethings.network/lorawan-stack/pkg/webui/console/api/index.js

  17:17  error  Unable to resolve path to module 'ttn-lw'  import/no-unresolved

✖ 1 problem (1 error, 0 warnings)

Error: running "node_modules/.bin/eslint ./pkg/webui/**/*.js ./config/**/*.js --no-ignore --color --quiet" failed with exit code 1

The command "if [[ "$RUNTYPE" == "js" ]]; then $MAGE js:lint jsSDK:lint; fi" exited with 1.
```

@kschiffer @bafonins can you guys shed some light on what's missing there for this to succeed?

**Comment by bafonins** [comment]
```
17:17  error  Unable to resolve path to module 'ttn-lw'  import/no-unresolved

✖ 1 problem (1 error, 0 warnings)
```

means that the javascript SDK (ttn-lw) which is a separate project with its own `package.json` is not linked into the main project. `JsSDK.Link` target is reposible for this.

**Comment by rvolosatovs** [comment]
> ```
> 17:17  error  Unable to resolve path to module 'ttn-lw'  import/no-unresolved
> 
> ✖ 1 problem (1 error, 0 warnings)
> ```
> 
> means that the javascript SDK (ttn-lw) which is a separate project with its own `package.json` is not linked into the main project. `JsSDK.Link` target is reposible for this.

Thanks for the response! I fixed it now (see CI)

**Comment by kschiffer** [comment]
Note that in production runs, the linking should actually not be performed. Instead, the actual local SDK module should be used:

```go
// Link links the local sdk package via `yarn link` to prevent caching issues.
func (k JsSDK) Link() error {
if nodeEnv := os.Getenv("NODE_ENV"); nodeEnv != "development" {
		if mg.Verbose() {
			fmt.Println("Linking not necessary when in production mode, exiting...")
		}
		return nil
	}
	…
}
```

It's then necessary to change the dependency order of the build step a little to ensure that the SDK is build before the JS modules are installed.

```go
// Build runs all necessary commands to build the console bundles and files.
func (js Js) Build() {
	mg.SerialDeps(JsSDK.Build, js.Deps, js.BuildDll, js.BuildMain)
}
```

If you want you can include this in this PR. I'm currently working on the GIthub Action setup for end-to-end tests, which will include this anyway.

**Comment by rvolosatovs** [comment]
> Note that in production runs (like in CI), the linking should actually not be performed. Instead, the actual local SDK module should be used:
> 
> ```go
> // Link links the local sdk package via `yarn link` to prevent caching issues.
> func (k JsSDK) Link() error {
> if nodeEnv := os.Getenv("NODE_ENV"); nodeEnv != "development" {
>         if mg.Verbose() {
>             fmt.Println("Linking not necessary when in production mode, exiting...")
>         }
>         return nil
>     }
>     …
> }
> ```

Added that

> It's then necessary to change the dependency order of the build step a little to ensure that the SDK is build before the JS modules are installed.
> 
> ```go
> // Build runs all necessary commands to build the console bundles and files.
> func (js Js) Build() {
>     mg.SerialDeps(JsSDK.Build, js.Deps, js.BuildDll, js.BuildMain)
> }
> ```
> 
> If you want you can include this in this PR. I'm currently working on the GIthub Action setup for end-to-end tests, which will include this anyway.

This should not be necessary - `js.BuildDll` and `js.BuildMain` should properly set dependencies, such that neither `SerialDeps`, nor `js.Deps` would be required.

**Comment by rvolosatovs** [comment]
Note, that result of this PR is that stuff like this just works™ (as it should have in the first place):
```
rm -rf node_modules doc/themes/the-things-stack/node_modules sdk/js/node_modules && mage js:build        
~/src/go.thethings.network/lorawan-stack ~/src/go.thethings.network/lorawan-stack
Skipping DLL module bundling (production mode)
+ yarn@1.22.0
added 1 package and audited 1 package in 0.46s
found 0 vulnerabilities

yarn install v1.22.0
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@1.2.12: The platform "linux" is incompatible with this module.
info "fsevents@1.2.12" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.1.2: The platform "linux" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.9: The platform "linux" is incompatible with this module.
info "fsevents@1.2.9" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning " > connected-react-router@6.5.2" has unmet peer dependency "react-router@^4.3.1 || ^5.0.0".
warning " > react-intl@3.1.11" has unmet peer dependency "@types/react@^16.0.0".
warning "@storybook/addon-actions > @storybook/api@5.3.18" has unmet peer dependency "regenerator-runtime@*".
[4/4] Building fresh packages...
Done in 15.88s.
yarn install v1.22.0
yarn run v1.22.0
$ node util/http-mapper.js
[1/4] Resolving packages...
⠁ File saved.
Done in 0.17s.
[2/4] Fetching packages...
info fsevents@1.2.12: The platform "linux" is incompatible with this module.
info "fsevents@1.2.12" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.1.2: The platform "linux" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...
Done in 3.52s.
yarn run v1.22.0
$ rm -rf dist; babel src -d dist
Successfully compiled 33 files with Babel.
Done in 3.05s.
mage js:build  240.16s user 25.21s system 179% cpu 2:27.74 total
```

**Comment by bafonins** [comment]
Fixed some issues.
It is worth testing:
1. in production mode
	* `mage js:build` - should properly prepare console/oauth production bundle
2. in development mode
	* `mage js:serve` - should setup `webpack-dev-server` (including reloading js sdk and updating intl messages)

I removed checks of modtime in `Link` target as the sdk needs to be linked only once. Also, I moved the `Link` target to js targets as it seems to be more appropriate place, since the main project (`package.json`) requires linking, a not the sdk as a standalone package.

After testing various use cases, it seems that `SerialDeps` is needed, since in some cases `mg.Deps` can cause `js.Deps` to be executed twice.

**Comment by rvolosatovs** [comment]
@kschiffer / @bafonins please test that everything works for you guys.
I tried `js:build`, `js:serve` and `js:lint` with different `NODE_ENV` values and everything seemed to work. CI passes as well.

File: tools/mage/js_sdk.go [hunk 1]
```diff
@@ -17,133 +17,105 @@ package ttnmage
...
-		}
-		yarn, err := k.yarn()
-		if err != nil {
-			return err
-		}
-
-		return yarn("run", "build")
+	ok, err := target.Dir(
+		filepath.Join("sdk", "js", "dist"),
+		filepath.Join("sdk", "js", "src"),
```
Comment (kschiffer) [inline tools/mage/js_sdk.go:61]:
```suggestion
		filepath.Join("sdk", "js", "src"),
		filepath.Join("sdk", "js", "generated")
```

File: tools/mage/js.go [hunk 2]
```diff
@@ -26,326 +26,293 @@ import (
...
-		return sh.RunV(nodeBin("yarn"), args...)
-	}, nil
+	return nil
 }
 
-func (js Js) node() (func(args ...string) error, error) {
-	return func(args ...string) error {
-		return sh.Run("node", args...)
-	}, nil
+func execYarn(stdout, stderr io.Writer, args ...string) error {
```
Comment (bafonins) [inline tools/mage/js.go:53]: is not used anywhere

File: tools/mage/js.go [hunk 3]
```diff
@@ -26,326 +26,293 @@ import (
...
+	if mg.Verbose() {
+		fmt.Println("Extracting frontend messages")
+	}
+	if err = sh.Rm(".cache/messages"); err != nil {
+		return err
+	}
+	if err = os.MkdirAll("pkg/webui/locales", 0755); err != nil {
+		return err
+	}
+	return js.runYarnCommand("babel", "pkg/webui")
```
Comment (bafonins) [inline tools/mage/js.go:210]: This forces `babel` to output all source to `stdout` (see travis log). 
Consider
```go
err = execYarn(nil, os.Stderr, "babel", "pkg/webui")
  if err != nil {
	return err
}
```

**Pull request merged**